app/javascript/src/controllers/score_controller.js

```diff
@@ -3,6 +3,29 @@ import $ from "jquery"
 
 const SCORES = [0, 1, 2, 3, 4, 5]
 
+/* This controller continuously validates all of the score entries in the
+ * assessment and goal setting page. It checks first that the score and goal
+ * are set to allowed values, and verifies that the score never exceeds the
+ * goal.
+ *
+ * A goals page can be of arbitrary length and has an arbitrary set of fields
+ * involved, so we cannot use the normal Stimulus targets as much as we
+ * normally would. Instead, the validation functions do a lot of traversing the
```

Comment:
The only DOM traversal that's done in the validation function is to find the "pair" of the element that is being changed. For example, if a score field is changed, we need to find the corresponding goal field to check that the score is less than the goal and vice versa.